python/ray/dag/compiled_dag_node.py

```diff
@@ -2095,89 +2112,270 @@ def visualize(
+                        is_multi_output[child] = True
+                        child2parent[child] = upstream_task_idx
+                    width_adjust = max(width_adjust, len(child_idx_list))
+
+            # Topological sort to determine layers
+            layers = defaultdict(list)
+            zero_indegree = deque(
+                [idx for idx in self.idx_to_task if indegree[idx] == 0]
+            )
+            layer_index = 0
+
+            while zero_indegree:
+                next_layer = deque()
+                while zero_indegree:
+                    task_idx = zero_indegree.popleft()
+                    layers[layer_index].append(task_idx)
+                    for downstream in adj_list[task_idx]:
+                        indegree[downstream] -= 1
+                        if indegree[downstream] == 0:
+                            next_layer.append(downstream)
+                zero_indegree = next_layer
+                layer_index += 1
+
+            # Print detailed node information
+            ascii_visualization += "Nodes Information:\n"
+            for idx, info in node_info.items():
+                ascii_visualization += f'{idx} [label="{info}"] \n'
+
+            # Print edges
+            ascii_visualization += "\nEdges Information:\n"
+            for upstream_task, downstream_task, type_hint in edge_info:
+                ascii_visualization += (
+                    f"{upstream_task} -> {downstream_task} [label={type_hint}]\n"
+                )
 
-            # Initialize the label and attributes
-            label = f"Task {idx}\n"
-            shape = "oval"  # Default shape
-            style = "filled"
-            fillcolor = ""
-
-            # Handle different types of dag_node
-            if isinstance(dag_node, InputNode):
-                label += "InputNode"
-                shape = "rectangle"
-                fillcolor = "lightblue"
-            elif isinstance(dag_node, InputAttributeNode):
-                label += f"InputAttributeNode[{dag_node.key}]"
-                shape = "rectangle"
-                fillcolor = "lightblue"
-            elif isinstance(dag_node, MultiOutputNode):
-                label += "MultiOutputNode"
-                shape = "rectangle"
-                fillcolor = "yellow"
-            elif isinstance(dag_node, ClassMethodNode):
-                if dag_node.is_class_method_call:
-                    # Class Method Call Node
-                    method_name = dag_node.get_method_name()
-                    actor_handle = dag_node._get_actor_handle()
-                    if actor_handle:
-                        actor_id = actor_handle._actor_id.hex()
-                        label += f"Actor: {actor_id[:6]}...\nMethod: {method_name}"
+            # Find the maximum width (number of nodes in any layer)
+            max_width = max(len(layer) for layer in layers.values()) + width_adjust
+            height = len(layers)
+
+            # Build grid for ASCII visualization
```

Comment:
Building this ourselves can be fragile. Can we utilize some ascii visualization library?

Reply:
Honestly, I have check some libs but they have different problems, Do you have any suggestions?
I have tried:
[etano/graphscii](https://github.com/etano/graphscii): This is not runnable
networkx: Not naively for ascii
https://github.com/cosminbasca/asciinet: Need additional package networkx 
https://github.com/sambrightman/asciidag: I dont think this is readable...

Reply:
I haven't researched into the ascii libraries. I'm OK using our own implementation if a good lib is not available.

It'd be nice if we know what kinds of DAGs the current implementation visualizes well (e.g., scale, shape etc.), and what are the limitations. Ideally this information can be included in the docstring so users have the right expectation.

Reply:
I just add doc about the algorithm and the limitation